java-checks/src/main/java/org/sonar/java/checks/SynchronizedClassUsageCheck.java

```diff
@@ -56,26 +56,37 @@
 
   private final Deque<Set<String>> exclusions = new ArrayDeque<>();
 
+  private final Set<Tree> visited = new HashSet<>();
+
   @Override
   public List<Tree.Kind> nodesToVisit() {
-    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
+    // We register on compilation units to clear the visited data structure when scanning a new file
+    return Arrays.asList(Tree.Kind.CLASS, Tree.Kind.COMPILATION_UNIT, Tree.Kind.ENUM, Tree.Kind.INTERFACE, Tree.Kind.ANNOTATION_TYPE);
   }
 
   @Override
   public void visitNode(Tree tree) {
+    if (tree.is(Tree.Kind.COMPILATION_UNIT)) {
+      // We clear the visited data structure when entering every file
+      visited.clear();
+      return;
+    }
     ExclusionsVisitor exclusionsVisitor = new ExclusionsVisitor();
     tree.accept(exclusionsVisitor);
     Set<String> currentClassExclusions = exclusionsVisitor.exclusions;
-    if(!exclusions.isEmpty()) {
+    if (!exclusions.isEmpty()) {
       currentClassExclusions.addAll(exclusions.peek());
     }
     exclusions.push(currentClassExclusions);
     tree.accept(new DeprecatedTypeVisitor());
   }
 
+
   @Override
   public void leaveNode(Tree tree) {
-    exclusions.pop();
+    if (!exclusions.isEmpty()) {
```

Comment:
nop, sorry I missed this.
`leaveNode()` is called as many times as `visitNode()`, and both methods are going to be called for all the kind we registered to.

For all the nodes but COMPILATION_UNIT, we push something in the stack, so for each of them we should pop(), and not pop() because it's not empty.

I would replace this by:
```java
if(!tree.is(Tree.Kind.COMPILATION_UNIT)) {
  exclusions.pop();
}